lib/pact/provider_verifier/add_interaction_attributes_header.rb

```diff
@@ -0,0 +1,39 @@
+require 'delegate'
+
+module Pact
+  module ProviderVerifier
+    class RequestDelegate < SimpleDelegator
+      def initialize request, extra_rack_headers
+        super(request)
+        @extra_rack_headers = extra_rack_headers
+      end
+
+      def headers
+        __getobj__().headers.merge(@extra_rack_headers)
+      end
+
+      def method
+        __getobj__().method
+      end
+    end
+
+    class AddInteractionAttributesHeader
+      def self.call(request, interaction)
+        extra_rack_headers = {
+          "X_PACT_DESCRIPTION" => interaction.description
+        }
+
+        if interaction.provider_states
+          extra_rack_headers["X_PACT_PROVIDER_STATES"] = interaction
+            .provider_states.collect(&:to_hash)
```

Comment:
This fails when parsing a v3 spec

```
header X-Gateway-Pact-States has field value "{\"name\"=>\"There is a greeting1\", \"params\"=>{\"id\"=>\"foo1\"}}\n{\"name\"=>\"There is a greeting2\", \"params\"=>{\"id\"=>2}}", this cannot include CR/LF
```

we would need to parse it into json, to pass it along the wire, I think

given this pact

```
{
  "consumer": {
    "name": "me"
  },
  "provider": {
    "name": "they"
  },
  "interactions": [
    {
      "description": "Provider state success",
      "providerState": "There is a greeting",
      "providerStates": [
        {
          "name": "There is a greeting1",
          "params": {
            "id": "foo1"
          }
        },
        {
          "name": "There is a greeting2",
          "params": {
            "id": 2
          }
        }
      ],
      "request": {
        "method": "GET",
        "path": "/somestate"
      },
      "response": {
        "status": 200,
        "headers": {},
        "body": {
          "greeting": "State data!"
        }
      }
    }
  ],
  "metadata": {
    "pactSpecificationVersion": "3.0.0"
  }
}
```